Implement osu!mania-specific health bar display

frenzibyte opened the pull request:
- Resolves https://github.com/ppy/osu/issues/22362

Added a migration step to move `LegacyHealthDisplay` from global target to per-ruleset target, as well as adding the mania-specific `LegacyHealthDisplay` specification for skins with existing user configuration. This required a minor refactor over the skin migration logic in general, to allow mutating multiple skin layouts in one step (especially when data from the first layout is required to act on the second layout).

frenzibyte committed:
Add osu!mania specifications for health bar display

frenzibyte committed:
Refactor skin migration to allow mutating multiple layouts at once

frenzibyte committed:
Migrate legacy health bar display to per-ruleset target

frenzibyte commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -293,20 +304,71 @@ private void applyMigration(SkinLayoutInfo layout, GlobalSkinnableContainers tar
+                        {
+                            if (ruleset.ShortName == @"mania")
+                            {
+                                // should avoid adding legacy health bar to non-legacy skins (unless explicitly added by user).
+                                if (!legacyHealthBars.Any())
+                                    break;
```
This implies that if the skin being migrated here has configured the `Playfield` target but never configured the non-ruleset `HUD` target, then the health bar will not be added.

It's a hard case to handle here because we don't really have a nice way of telling whether a skin is based on `LegacySkin`, nor do I want to use that as a condition anywhere anyway. I think it's fine to let those kind of skins manually add their health bar by resetting the target or otherwise.

bdach replied:
I've read this comment five times and I still can't understand what you're trying to say. Can you elaborate? Doesn't the HUD target read fall back to an empty array if it fails? Can't you use that?

frenzibyte replied:
> Doesn't the HUD target read fall back to an empty array if it fails?

What does "fails" mean here.

Let me attempt to explain how the migration logic works in a case-by-case scenario:
 - A skin with no configuration
   - `HUD` and `Playfield` are all null, so the lookups will get the default components from `LegacySkin` and `ManiaLegacySkinTransformer` if it's a legacy skin, we're gucci.
 - A skin with `HUD` configuration but no `Playfield` configuration
   - `HUD` is not null and `Playfield` is null
   - If the `HUD` configuration has a legacy health bar, remove it and store it to `legacyHealthBars`.
   - Don't do anything for `Playfield`, it's already null, the health bar will be looked up from `ManiaLegacySkinTransformer` if it's a legacy skin.
 - A skin with `HUD` configuration and `Playfield` configuration
   - `HUD` and `Playfield` are both not null
   - If the `HUD` configuration has a legacy health bar, remove it and store it to `legacyHealthBars`.
   - If `legacyHealthBars` is not empty, add a legacy mania health bar to `Playfield`.
     - The condition used here is good enough to ensure if the configured skin is not legacy then no legacy health bar will be added to mania, by checking whether the skin used to have a legacy health bar or not.
 - A skin with no `HUD` configuration but a `Playfield` configuration
   - `HUD` is null, `Playfield` is not null
   - Don't do anything for `HUD`, it's already null
   - `legacyHealthBars` is empty because `HUD` is null
   - Don't do anything for `Playfield` since `legacyHealthBars` is empty
     - This is the part touched on in this conversation. This is incorrect because if the skin is legacy then there should be a health bar added to `Playfield` by this migration logic, but I'm intentionally ignoring it for simplicity. It should be a rare case and one that can be fixed by the user.

frenzibyte replied:
Might as well write test cases rather than state them word by word.

frenzibyte replied:
Test cases written within an introduced testing system `TestSceneSkinMigration`.

bdach commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -148,6 +149,28 @@ protected Skin(SkinInfo skin, IStorageResourceProvider? resources, IResourceStor
+
+            if (layoutInfos.Values.Any())
+            {
+                int version = layoutInfos.Values.Min(l => l.Version);
+
+                for (int i = version + 1; i <= SkinLayoutInfo.LATEST_VERSION; i++)
+                    applyMigration(i);
```
There should be no taking of `.Min()` here, and `applyMigration()` should be called in the inner loop over `layoutInfos.Values`.

Why? Presume you have two `LayoutInfo`s with different versions each. `LayoutInfo` X is on version 1, and `LayoutInfo` Y on version 2. Now when the latest version is bumped to 3, _Y will have the migration from version 1 to 2 called twice_. Which is _only fine_ if the migration is _idempotent_, as in it can _determine whether it has already run, and bail if it has_.

This is a _very not obvious_ property of a migration system; with these you generally assume that a migration is a _one-and-done_ affair. Thus I object to this type of handling, as an unnecessary complication. If targets are allowed to deviate in version, then migrations should be applied on a per-target basis.

frenzibyte replied:
> and `applyMigration()` should be called in the inner loop over layoutInfos.Values.

Read OP.

The intention is that we should never have a case where there are two layouts with different migration versions, unless the user decided to mess around with the files. And, hopefully our migration system is good enough to not affect layouts which are already migrated even if the migration was ran again, but that's just a loose 
assumption.

That being said, with the concern raised both here and essentially https://github.com/ppy/osu/pull/30867#discussion_r1856466286, I will revert this and compare against the instantiation type instead. I realised at the end of the day if a configuration is null then the default components technically rely on the "instantiation type" (what type of skin class the skin is), so it's more logical to read the instantiation type and go from there instead.

bdach replied:
> The intention is that we should never have a case where there are two layouts with different migration versions, unless the user decided to mess around with the files.

You can't assume that people don't mess with the files, because they already do, and will become to do it more after external skin mounting is added.

bdach commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -293,20 +304,71 @@ private void applyMigration(SkinLayoutInfo layout, GlobalSkinnableContainers tar
+                {
+                    // Legacy health bar is moved from the global HUD components target into per-ruleset.
+                    // On osu!mania, the health bar is moved from the HUD to Playfield target for stage-based positioning.
+                    if (resources == null)
+                        break;
```
What is this check... checking for? It's completely inscrutable to me.

frenzibyte replied:
Mostly ignore doing anything if we don't have the thing to do it, it only matters in tests where skins are instantiated without resources. This check existed before so I'm not sure why it's being asked in here.

bdach commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -293,20 +304,71 @@ private void applyMigration(SkinLayoutInfo layout, GlobalSkinnableContainers tar
+                    if (hudLayout == null || !hudLayout.TryGetDrawableInfo(null, out var globalHUDComponents))
+                        globalHUDComponents = Array.Empty<SerialisedDrawableInfo>();
+
+                    var legacyHealthBars = globalHUDComponents.Where(h => h.Type.Name == nameof(LegacyHealthDisplay)).ToArray();
```
Why is this checking the _name_ of the type _if you already have the type itself in hand_?

```suggestion
                    var legacyHealthBars = globalHUDComponents.Where(h => h.Type == typeof(LegacyHealthDisplay)).ToArray();
```

frenzibyte replied:
Sure, I don't know how `Type` equality works, and I hope that if it does other stupid things like assembly version or whatever then it won't break this logic.

frenzibyte replied:
I will assume it only checks the name. I tried stepping through the equality implementation and ended up with both sides being equal by reference, I don't know.

peppy replied:
If you need clarification here, given that `CreateInstance` works with however we're serialising types, this case also has to work. We do seem to be serialising versioning out, but I guess it works fine.

frenzibyte committed:
Revert "Migrate legacy health bar display to per-ruleset target"

This reverts commit 76f79ce083a6adfee7fd2e23dbb5be2ffb4ac947.

frenzibyte committed:
Revert "Refactor skin migration to allow mutating multiple layouts at once"

This reverts commit f1b56869042ef2cd3ac8174c7d8a7c252247b9a2.

frenzibyte committed:
Add extensive test scene for skin migrations

This is different from `SkinDeserialisationTest` in that layouts can be written programmatically with as much ease, allowing to test migration logic with different scenarios without running the game and exporting skins and attaching them to tests.

frenzibyte committed:
Add migration logic for legacy health displays

frenzibyte committed:
Move test and mark headless

frenzibyte committed:
Use type equality

frenzibyte commented:
I'll update the deserialisation test once the PR is considered ready to merge.

peppy commented on osu.Game.Tests/Skins/TestSceneSkinMigration.cs:
```diff
@@ -0,0 +1,407 @@
+                });
+            });
+
+            // One may argue that if a LegacyHealthDisplay exists in a non-legacy skin,
```
I'd prefer these comments are in the actual migration rather than the test.

ie. for this one it should be at https://github.com/ppy/osu/pull/30867/files#diff-e662eed704866ec7b4c4ec234173d642e8628330ab7b5c4480e7f69070ca0e2dR345.

frenzibyte replied:
I've moved the one highlighted above, but there's also this:

https://github.com/ppy/osu/blob/5ab9074f5dbd637eac803cf87f1b64aeb2083c48/osu.Game.Tests/Skins/TestSceneSkinMigration.cs#L247-L248 

And there's no direct place to move it to the migration code, and I intended it to be in the test specifically to explain why the test case is written as such.

frenzibyte committed:
Move non-legacy conditional reasoning to migration code

peppy commented:
@frenzibyte  Probably safe to update the failing migration tests now, I'm okay with how this is looking.

frenzibyte committed:
Refactor skin deserialisation tests to better support skin migration

frenzibyte commented:
I thought the skin deserialisation test failure was happening due to missing types or something, but turns out it was because of addressing https://github.com/ppy/osu/pull/30867#discussion_r1856454845. It appears `SkinDeserialisationTest` was never able to run migrations, and the test results on some tests were outdated due to not being migrated.

I don't think this is correct behaviour, it feels quite trippy for a test to load an osk file and not have it migrated. I've rewritten the test to import the skin to a `SkinManager` instead.

frenzibyte commented on osu.Game.Tests/Skins/TestSceneSkinDeserialisation.cs:
```diff
@@ -0,0 +1,194 @@
+            AddStep("load skin", () => skin = importSkinFromArchives("Archives/modified-argon-20231106.osk"));
+
+            AddAssert("layouts count = 2", () => skin.LayoutInfos, () => Has.Count.EqualTo(2));
+            AddAssert("hud count = 13",
+                () => skin.LayoutInfos[GlobalSkinnableContainers.MainHUDComponents].AllDrawables.ToArray(),
+                () => Has.Length.EqualTo(13));
```
Some may notice the count assertions on some of these tests have changed. This is because previously the test did not count the combo counter migration, which moves the combo counter from the global layout to each ruleset's layout, therefore increasing the `AllDrawables` count by 3.

bdach committed:
Document & reduce visibility of `GetInstanceType()`

bdach reviewed (changes requested):
Source looks fine (aside from https://github.com/ppy/osu/pull/30867/commits/c12cb2582e2ff15b015d21e95b384623c00237b3 that I pushed in because the method had a bad name). *However*, for whatever reason the legacy default skin does not have a health bar in osu! and osu!catch *specifically* (taiko and mania somehow work):

https://github.com/user-attachments/assets/aa2b4933-336e-45b6-951c-0c3958812628

Not sure why this passes tests, but I'm not particularly wanting to investigate this, so leaving it to you to diagnose.

frenzibyte commented:
Default ruleset-based implementations of osu! and osu!catch rulesets override the health display set up by LegacySkin... I'll need to investigate how to add appropriate tests for presence of default components.

frenzibyte committed:
Merge branch 'master' into mania-hp-bar-position

frenzibyte committed:
Mark skin deserialisation test as headless

frenzibyte committed:
Add test scene for checking all displayed default skin components per screen

frenzibyte committed:
Fix osu! and osu!catch no longer displaying a health display

frenzibyte commented:
Fortunately enough the fix was simple. Added tests to ensure this doesn't slip by, by being critical of every default skinnable component displayed per skin. That area was not being appropriately tested by any of our existing tests so I rolled out a test scene for it, counting a total of 2 new test scenes and a rewrite of an existing one :^).

frenzibyte commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -364,6 +364,8 @@ protected override void ParseConfigurationStream(Stream stream)
                         case GlobalSkinnableContainers.MainHUDComponents:
                             if (containerLookup.Ruleset != null)
                             {
+                                // Notably, there may be existing ruleset skin transformers that override the default layout returned here.
+                                // Any added component in here should be considered in the existing transformers if applicable.
```
Added comment because I tripped on this one and caused a total disaster (as seen above).

frenzibyte committed:
Remove redundant array creation expressions

Okay.

bdach commented:
I'm not sure how much of a big deal to make of this but this diff can seemingly cause health bars to be doubled up on skins that already have a health bar in the playfield target.

Check out what happens when you import https://osu.ppy.sh/community/forums/topics/1923469?n=1 on master and switch to this branch. For me this happens:

![osu_2024-12-11_17-18-11](https://github.com/user-attachments/assets/dfed1d45-abcf-4a6e-bbce-b8b425c3e912)
![osu_2024-12-11_17-16-00](https://github.com/user-attachments/assets/1670cf9a-e5c0-432d-b2c3-4e23aa078951)

Seems like this'll be the migration doing undesirable things...

frenzibyte commented:
I think it's fine to exclude skins which already have a health bar on the playfield target.

frenzibyte committed:
Add failing test case

frenzibyte committed:
Exclude skins that contain health display on playfield target

frenzibyte committed:
Merge branch 'master' into mania-hp-bar-position

bdach reviewed (dismissed):
seems probably ok although I will complain that this has ballooned to ~~nearly~~ over 1k lines of diff

@peppy not sure if you want to cross-check this or not

peppy commented:
I was going to get this in but the deserialisation test conflict is a bit of a pain to fix up.

frenzibyte committed:
Merge branch 'master' into mania-hp-bar-position

peppy committed:
Merge branch 'master' into mania-hp-bar-position

peppy commented on osu.Game/Skinning/SkinInfo.cs:
```diff
@@ -52,13 +55,15 @@ public virtual Skin CreateInstance(IStorageResourceProvider resources)
-        public IList<RealmNamedFileUsage> Files { get; } = null!;
+        public virtual Skin CreateInstance(IStorageResourceProvider resources) => (Skin)Activator.CreateInstance(GetInstanceType(), this, resources)!;
+
+        public virtual IList<RealmNamedFileUsage> Files { get; } = null!;
```
I'm surprised realm even allows you to make this `virtual`.... Seems like a very wrong direction.

Let's not, unless there's another case we've done this.

frenzibyte replied:
Seems like I don't even need it. The usage was inserting dummy realm files in `TestSceneSkinMigration.TestSkinInfo`, and removing the override does not seem to have any effect on the test results.

peppy commented on osu.Game/Skinning/SkinInfo.cs:
```diff
@@ -39,7 +39,10 @@ public class SkinInfo : RealmObject, IHasRealmFiles, IEquatable<SkinInfo>, IHasG
 
-        public virtual Skin CreateInstance(IStorageResourceProvider resources)
+        /// <summary>
+        /// Returns the specific subtype of <see cref="Skin"/> that will be constructed on calling <see cref="CreateInstance"/>.
```
Except it returned `TrianglesSkin` in a case where it's no longer returned in `CreateInstance`? This change makes zero sense when only looking at the diff, so I'm going to require further explanation.

frenzibyte replied:
I don't understand your concern. This is just splitting what was in `CreateInstance` to its own method, I'm not sure what's wrong here.

frenzibyte committed:
Remove unnecessary override and weird usage of Realm
